File: src/paste.ts
import { captureClipboardImage } from "./clipboard.js";
import { AppConfig } from "./config.js";
import { resolveTargetPane } from "./panes.js";
import { uploadImage } from "./remote-cache.js";
import { runRemoteHelper } from "./remote-helper.js";

export async function pasteClipboardImage(config: AppConfig, targetPane?: string): Promise<void> {
  const image = await captureClipboardImage();
  const remotePath = await uploadImage(config, image);
  const target = targetPane ?? (await resolveTargetPane(config));
  await injectText(config, target, remotePath);
  console.log(`pasted ${remotePath} into ${target}`);
}

export async function injectText(config: AppConfig, targetPane: string, text: string): Promise<void> {
  const result = await runRemoteHelper(config, ["inject", targetPane], text);
  if (result.exitCode !== 0) throw new Error(`Remote paste failed: ${result.stderr.trim() || result.stdout.trim()}`);
}

File: src/paste.rs
use anyhow::{Result, bail};

use crate::clipboard::capture_clipboard_payload;
use crate::config::AppConfig;
use crate::panes::resolve_target_pane;
use crate::remote_cache::upload_payload;
use crate::remote_helper::run_remote_helper;

pub fn paste_clipboard_payload(config: &AppConfig, target_pane: Option<&str>) -> Result<()> {
    let payload = capture_clipboard_payload()?;
    let remote_path = upload_payload(config, &payload)?;
    let target = match target_pane {
        Some(target) => target.to_string(),
        None => resolve_target_pane(config)?,
    };
    inject_text(config, &target, &remote_path)?;
    println!("pasted {remote_path} into {target}");
    Ok(())
}

pub fn inject_text(config: &AppConfig, target_pane: &str, text: &str) -> Result<()> {
    let result = run_remote_helper(
        config,
        &["inject".to_string(), target_pane.to_string()],
        Some(text.as_bytes()),
    )?;
    if result.exit_code != 0 {
        bail!(
            "remote paste failed: {}",
            if result.stderr.trim().is_empty() {
                result.stdout.trim()
            } else {
                result.stderr.trim()
            }
        );
    }
    Ok(())
}
